File: src/main/java/com/jhl/mds/dto/TaskDTO.java
package com.jhl.mds.dto;

import com.jhl.mds.consts.MySQLInsertMode;
import com.jhl.mds.consts.TaskType;
import com.jhl.mds.dao.entities.Task;
import com.jhl.mds.dao.entities.TaskFieldMapping;
import com.jhl.mds.dao.entities.TaskFilter;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TaskDTO {
    private int taskId;
    private String taskName;
    private List<SimpleFieldMappingDTO> mapping;
    private Table source;
    private Table target;
    private TaskType taskType;
    private MySQLInsertMode insertMode;
    private List<String> filters;

    @Data
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Table {
        private int serverId;
        private String database;
        private String table;
    }

    @Component
    public static class Converter {

        public TaskDTO from(Task task, List<TaskFieldMapping> taskFieldMappings, List<TaskFilter> taskFilters) {
            TaskDTOBuilder builder = TaskDTO.builder()
                    .taskId(task.getTaskId())
                    .taskName(task.getName())
                    .source(Table.builder()
                            .serverId(task.getFkSourceServer())
                            .database(task.getSourceDatabase())
                            .table(task.getSourceTable())
                            .build())
                    .target(Table.builder()
                            .serverId(task.getFkTargetServer())
                            .database(task.getTargetDatabase())
                            .table(task.getTargetTable())
                            .build())
                    .taskType(TaskType.getByCode(task.getTaskType()))
                    .insertMode(MySQLInsertMode.valueOf(task.getInsertMode()));

            if (taskFieldMappings != null) {
                List<SimpleFieldMappingDTO> mapping = taskFieldMappings.stream()
                        .map(o -> new SimpleFieldMappingDTO(o.getSourceField(), o.getTargetField())).collect(Collectors.toList());
                builder.mapping(mapping);
            }

            if (taskFilters != null) {
                List<String> filters = taskFilters.stream().map(TaskFilter::getFilter).collect(Collectors.toList());
                builder.filters(filters);
            }

            return builder.build();
        }

        public TaskDTO from(Task task) {
            return from(task, null, null);
        }
    }
}

File: src/main/resources/static/resources/js/source/dto/task-dto.js
// @flow
import {MySQLInsertMode} from './my-sqlinsert-mode';
import {SimpleFieldMappingDTO} from './simple-field-mapping-dto';
import {TaskType} from './task-type';
import {Table} from './table';
import {observable} from 'mobx';

export class TaskDTO {

    /**
     * @type {number}
     */
    @observable taskId: ?number = 0;
    /**
     * @type {string}
     */
    @observable taskName: ?string = '';
    /**
     * @type {SimpleFieldMappingDTO[]}
     */
    @observable mapping: ?SimpleFieldMappingDTO[] = null;
    /**
     * @type {Table}
     */
    @observable source: ?Table = null;
    /**
     * @type {Table}
     */
    @observable target: ?Table = null;
    /**
     * @type {TaskType}
     */
    @observable taskType: ?TaskType = null;
    /**
     * @type {MySQLInsertMode}
     */
    @observable insertMode: ?MySQLInsertMode = null;
    /**
     * @type {string[]}
     */
    @observable filters: ?string[] = null;

    /**
     * @param taskId {number}
     * @param taskName {string}
     * @param mapping {SimpleFieldMappingDTO[]}
     * @param source {Table}
     * @param target {Table}
     * @param taskType {TaskType}
     * @param insertMode {MySQLInsertMode}
     * @param filters {string[]}
     */
    constructor(taskId: ?number, taskName: ?string, mapping: ?SimpleFieldMappingDTO[], source: ?Table, target: ?Table, taskType: ?TaskType, insertMode: ?MySQLInsertMode, filters: ?string[]) {
        this.taskId = taskId;
        this.taskName = taskName;
        this.mapping = mapping;
        this.source = source;
        this.target = target;
        this.taskType = taskType;
        this.insertMode = insertMode;
        this.filters = filters;
    }


}
